fvm/environment/program_recovery.go

```diff
@@ -0,0 +1,203 @@
+
+                  access(all)
+                  var balance: UFix64
+
+                  init(balance: UFix64) {
+                      self.balance = balance
+                  }
+
+                  access(FungibleToken.Withdraw)
+                  fun withdraw(amount: UFix64): @{FungibleToken.Vault} {
+                      panic("withdraw is not implemented")
+                  }
+
+                  access(all)
+                  view fun isAvailableToWithdraw(amount: UFix64): Bool {
+                      panic("isAvailableToWithdraw is not implemented")
+                  }
+
+                  access(all)
+                  fun deposit(from: @{FungibleToken.Vault}) {
+                      panic("deposit is not implemented")
+                  }
+
+                  access(all)
+                  fun createEmptyVault(): @{FungibleToken.Vault} {
+                      panic("createEmptyVault is not implemented")
+                  }
+              }
+
+              access(all)
+              fun createEmptyVault(vaultType: Type): @{FungibleToken.Vault} {
+                  panic("createEmptyVault is not implemented")
+              }
+          }
+        `,
+		fungibleTokenAddress.HexWithPrefix(),
+		addressLocation.Name,
+	)
+
+	return parser.ParseProgram(memoryGauge, []byte(code), parser.Config{})
+}
+
+func importsAddressLocation(program *ast.Program, address common.Address, name string) bool {
+	importDeclarations := program.ImportDeclarations()
+
+	// Check if the location is imported by any import declaration
+	for _, importDeclaration := range importDeclarations {
+
+		// The import declaration imports from the same address
+		importedLocation, ok := importDeclaration.Location.(common.AddressLocation)
+		if !ok || importedLocation.Address != address {
+			continue
+		}
+
+		// The import declaration imports all identifiers, so also the location
+		if len(importDeclaration.Identifiers) == 0 {
+			return true
+		}
+
+		// The import declaration imports specific identifiers, so check if the location is imported
+		for _, identifier := range importDeclaration.Identifiers {
+			if identifier.Identifier == name {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func declaresConformanceTo(conformingDeclaration ast.ConformingDeclaration, name string) bool {
+	for _, conformance := range conformingDeclaration.ConformanceList() {
+		if conformance.Identifier.Identifier == name {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isNominalType(ty ast.Type, name string) bool {
+	nominalType, ok := ty.(*ast.NominalType)
+	return ok &&
+		len(nominalType.NestedIdentifiers) == 0 &&
+		nominalType.Identifier.Identifier == name
+}
+
+const fungibleTokenTypeTotalSupplyFieldName = "totalSupply"
+const fungibleTokenVaultTypeBalanceFieldName = "balance"
+const fungibleTokenVaultTypeIdentifier = "Vault"
+
+func isFungibleTokenContract(program *ast.Program, fungibleTokenAddress common.Address) bool {
+	const contractName = "FungibleToken"
```

Comment:
a nitpick: maybe rename this variable to `fungibleTokenContract`/ `fungibleTokenContractName`, so the code below reads a bit better. e.g: `declaresConformanceTo(contractDeclaration, fungibleTokenContract)`

Reply:
Good idea! Done in [6c46cee](https://github.com/onflow/flow-go/pull/6278/commits/6c46ceef0e63a9e5d6ddff8bf7bdd26cfa209298)